File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode temp = head;
        ListNode bye = head;
        ListNode prev = null;

        while(n > 1){
            temp = temp.next;
            n--;
        }

        while(temp.next != null) {
            prev = bye;
            bye = bye.next;
            temp = temp.next;
        }

        if(prev != null)
            prev.next = bye.next;
        else
            head = head.next;

        return head;
    }
}

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dum = ListNode(-1,head)
        end = head
        prev = dum
        while n > 0 and end:
            end = end.next
            n -= 1

        while end:
            end = end.next
            prev = prev.next

        if prev.val == -1:
            return head.next
        else:
            prev.next = prev.next.next

        return head